netx/net/sourceforge/jnlp/util/XDesktopEntry.java

```diff
@@ -477,13 +485,98 @@ private void cacheIcon() throws IOException, NonFileProtocolException {
+        try {
+            BufferedImage img = ImageIO.read(source);
+            short bitCount = (short)img.getColorModel().getPixelSize();
+            // Images with less than 32 bit color depth 
+            // are very hard to produce in ico format.
+            // Therefore, the image is converted if necessary
+            if (bitCount < 32)
+            {
+                BufferedImage dest = new BufferedImage(
+      			  	img.getWidth(), img.getHeight(),
+        			BufferedImage.TYPE_INT_ARGB);
+                ColorConvertOp cco = new ColorConvertOp(
+        			img.getColorModel().getColorSpace(),
+        			dest.getColorModel().getColorSpace(),
+        			null);
+                img = cco.filter(img, dest);
+                bitCount = (short)img.getColorModel().getPixelSize();
+            }
+            ByteArrayOutputStream bos = new ByteArrayOutputStream();
+            boolean written = ImageIO.write(img, "png", bos);
+            if (!written)
+                return null;
+                        
+            byte width  = (byte)(img.getWidth()  < 256 ? img.getWidth()   : 0);
+            byte height = (byte)(img.getHeight() < 256 ? img.getHeight()  : 0);
+            byte colorCount = (byte)(bitCount < 8 ? Math.pow(2, bitCount) : 0);            
+            byte[] imgBytes = bos.toByteArray();
+            int offset = 22;
+            int fileSize = imgBytes.length + offset;
+            ByteBuffer bytes = ByteBuffer.allocate(fileSize);
+            bytes.order(ByteOrder.LITTLE_ENDIAN);
+            
+            // Header
+            bytes.putShort((short) 0);      // Reserved, must be 0
+            bytes.putShort((short) 1);      // Image type: 1 for ico
+            bytes.putShort((short) 1);      // Number of images in the file
+            // Directory Entry
+            bytes.put(width);               // Image width
+            bytes.put(height);              // Image height
+            bytes.put(colorCount);          // Number of colors
+            bytes.put((byte) 0);            // Reserved, must be 0
+            bytes.putShort((short) 1);      // Number of color planes
+            bytes.putShort(bitCount);       // Number of bits per pixel
+            bytes.putInt(imgBytes.length);  // Image size
+            bytes.putInt(offset);           // Image offset
+            // Image
+            bytes.put(imgBytes);            // Image data
+
+            File target = new File(PathsAndFiles.ICONS_DIR.getFile(), 
+                                   targetName.substring(0, 
+                                   targetName.lastIndexOf('.')) + ".ico"); 
+            try (FileOutputStream fos = new FileOutputStream(target)) {
+                fos.write(bytes.array());
+                fos.flush();
+            }
+            return target;
+        }
+        catch (IOException e) {
+            return null;
+        }
+    }
+    
+    String cacheAndGetIconLocation() {        
```

Comment:
add `private`

Reply:
This method is used by WindowsDesktopEntry and therefore can not be private.

Reply:
than protected

Reply:
package-private is stricter than protected